### PR TITLE
[SPARK-18511] Added an api for join operations just on column name and the join type

### PR DESCRIPTION
**Pull request by shiv4nsh**
Added an api for join operations just on column name and the join type. An Api in which we do not have to provide the column sequences.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -705,12 +701,10 @@ class Dataset[T] private[sql](
...
-   *
-   * @note If you perform a self-join using this function without aliasing the input
-   * [[DataFrame]]s, you will NOT be able to reference any columns after the join, since
-   * there is no way to disambiguate which side of the join you would like to reference.
-   *
-   * @group untypedrel
+    * @note If you perform a self-join using this function without aliasing the input
+   *        [[DataFrame]]s, you will NOT be able to reference any columns after the join, since
+   *        there is no way to disambiguate which side of the join you would like to reference.
+    * @group untypedrel
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Can we exclude unrelated changes here? Also, `@note` has the consistent indentation across documentation. I guess it is not encouraged to change some of them inconsistently without a concrete reason.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Oh, no. I mean some changes here such as

```diff
-   *
-   * @group action
+    * @group action
```

seems not related with what the PR tries to proposes.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: It seems my IntelliJ does not. Whether it is a problem or not, reformatting seems not the goal of this PR and I guess it seems reformatting is arguable.

**Comment by shiv4nsh** [comment]
@HyukjinKwon : I have Closed this PR and added a new PR #15944 with the main changes that were suggested in this PR .
